src/Microsoft.VisualStudio.Editor.Razor/BraceSmartIndenter.cs

```diff
@@ -0,0 +1,218 @@
+                // Save the caret position
+                var textView = context.FocusedTextView;
+                var caret = textView.Caret.Position.BufferPosition;
+
+                // Insert the completion text, i.e. "\r\n      "
+                var edit = textView.TextBuffer.CreateEdit();
+                var indent = CalculateIndent(textView.TextBuffer, context.ChangePosition);
+                edit.Insert(caret.Position, indent);
+                edit.Apply();
+
+                // Place the caret inbetween the braces (before our indent).
+                var currentSnapshotPoint = new SnapshotPoint(textView.TextBuffer.CurrentSnapshot, caret.Position);
+                textView.Caret.MoveTo(currentSnapshotPoint);
+
+                // For Razor metacode cases the editors smart indent wont kick in automatically. This forces the smart indent.
+                // For example attempting to do enter a newline between the functions directive: @functions {} will not auto-indent
+                // inbetween the braces unless we forcefully move to end of line.
+                var editorOperations = _editorOperationsFactory.GetEditorOperations(textView);
+                editorOperations.MoveToEndOfLine(false);
+            }
+        }
+
+        // Internal for testing
+        internal void TextBuffer_OnChanged(object sender, TextContentChangedEventArgs contentChange)
+        {
+            _dispatcher.AssertForegroundThread();
+
+            if (contentChange.Changes.Count == 0)
+            {
+                return;
+            }
+
+            var firstChange = contentChange.Changes[0];
+            var lastChange = contentChange.Changes[contentChange.Changes.Count - 1];
+            var oldLength = lastChange.OldEnd - firstChange.OldPosition;
+            var newLength = lastChange.NewEnd - firstChange.NewPosition;
+            var newText = contentChange.After.GetText(firstChange.NewPosition, newLength);
+
+            var wasChanged = true;
+            if (oldLength == newLength)
+            {
+                var oldText = contentChange.Before.GetText(firstChange.OldPosition, oldLength);
+                wasChanged = !string.Equals(oldText, newText, StringComparison.Ordinal);
+            }
+
+            if (!wasChanged)
+            {
+                return;
+            }
+
+            if (TryCreateIndentationContext(firstChange.NewPosition, newLength, newText, _textBuffer, out var context))
+            {
+                _context = context;
+            }
+        }
+
+        // Internal for testing
+        internal static bool TryCreateIndentationContext(int changePosition, int changeLength, string finalText, ITextBuffer textBuffer, out BraceIndentationContext context)
+        {
+            var focusedTextView = textBuffer.GetFocusedTextView();
+            if (focusedTextView != null && ParserHelpers.IsNewLine(finalText))
+            {
+                var currentSnapshot = textBuffer.CurrentSnapshot;
```

Comment:
I don't feel like I super understand what this code is doing

Reply:
I'll pull pieces out into methods to be more descriptive.